JavaScript/Closures.md

```diff
@@ -0,0 +1,174 @@
+
+With a small tweak, we can make `inner()` available to the global space without changing the scope of the variables involved:
+
+```javascript
+var a = 1;
+function outer(){
+  var b = 2;
+  var inner = function(){
+    var c = 3;
+    return a + b + c;
+  }
+  return inner;
+}
+
+var d = outer();
+
+b; // undefined
+d(); // 6
+```
+
+Here, we've replaced the function declaration of `inner()` with a function expression, and then returned the function stored in `inner`, rather than its return value.
+
+When we assign the function to `d`, calling `d()` from the global space has the same effect as calling `inner()`, except this time we get the returned value of `6`, instead of an error. If we try to access `b` from the same space, we still get an error like last time.
+
+This is unusal because now `d()` has access to all 3 variables `a`, `b` and `c` in the global space, but any new function will not. Although it is defined in the global space, `d()` has **'remembered' the original environment (scope) in which it was created**.
+
+---- 
+
+### Why... do I need to know this?
+
+#### Avoid exposing data you don't need to
+
+Closures are useful when you want to set a variable to use in a function that shouldn't be accessed or changed by other functions, either intentionally or accidentally.
+
+Rather than using an object with properties and methods, you can use a closure to make sure that the properties don't get altered.
+
+#### Object
+
+```javascript
+var human = {
+  greeting : 'Hello!',
+  sayHello : function(){
+               console.log(this.greeting);
+             }
+
+human.greeting = 'Go away.';
+human.sayHello(); // outputs 'Go away.' :( 
+```
+
+#### Closure (with [immediately invoked function expression][3])
+
+```javascript
+var sayHello = (function(){
+  var greeting = 'Hello!';
+  var greet = function(){
+    console.log(greeting);
+  }
+  return greet;
+})();
+
+var greeting = 'Go away.';
+sayHello(); // outputs 'Hello!' :)
+```
+
+---- 
+
+#### Maintaining state
+  
+In the code below, the closure sets up an `i` value of `0` and updates the value every time the function is called. This allows iteration through an array one element at a time - the function 'remembers' the state of `i` at the last iteration.
+
+```javascript
+var array = ['a','b','c','d','e','f','g'];
+
+function iterate(arr){
+  var i = 0;
+  var iterator = function(){ 
+    return arr[i++];
+  }
+  return iterator;
+};
+
+var next = iterate(array);
+
+next(); // returns 'a'
+next(); // returns 'b'
+next(); // returns 'c'
+next(); // returns 'd'
+...
+```
+
+---- 
+
+### Related
+
+[JavaScript Scope: A Beginners Perspective][4]  
+[Variable scope and hoisting][5]
+[Hoisting][6]
+
```

Comment:
I wrote an article on closures in today I learned https://github.com/codingforeveryone/today-i-learned/blob/master/oojs/closures.md